common/management/src/role/role_mgr.rs

```diff
@@ -0,0 +1,191 @@
+        let value = serde_json::to_vec(&role_info)?;
+
+        let kv_api = self.kv_api.clone();
+        let upsert_kv = kv_api.upsert_kv(UpsertKVAction::new(
+            &key,
+            match_seq,
+            Operation::Update(value),
+            None,
+        ));
+        let res = upsert_kv.await?.into_add_result()?;
+        match res.res {
+            OkOrExist::Ok(v) => Ok(v.seq),
+            OkOrExist::Exists(v) => Err(ErrorCode::UserAlreadyExists(format!(
+                "Role already exists, seq [{}]",
+                v.seq
+            ))),
+        }
+    }
+
+    async fn get_role(&self, role_name: &str, seq: Option<u64>) -> Result<SeqV<RoleInfo>> {
+        let key = format!("{}/{}", self.role_prefix, role_name);
+        let kv_api = self.kv_api.clone();
+        let res = kv_api.get_kv(&key).await?;
+        let seq_value =
+            res.ok_or_else(|| ErrorCode::UnknownRole(format!("unknown role {}", role_name)))?;
+
+        match MatchSeq::from(seq).match_seq(&seq_value) {
+            Ok(_) => Ok(seq_value.into_seqv()?),
+            Err(_) => Err(ErrorCode::UnknownRole(format!(
+                "unknown role {}",
+                role_name
+            ))),
+        }
+    }
+
+    async fn get_roles(&self) -> Result<Vec<SeqV<RoleInfo>>> {
+        let role_prefix = self.role_prefix.clone();
+        let kv_api = self.kv_api.clone();
+        let values = kv_api.prefix_list_kv(role_prefix.as_str()).await?;
+
+        let mut r = vec![];
+        for (_key, val) in values {
+            let u = serde_json::from_slice::<RoleInfo>(&val.data)
+                .map_err_to_code(ErrorCode::IllegalUserInfoFormat, || "")?;
+
+            r.push(SeqV::new(val.seq, u));
+        }
+
+        Ok(r)
+    }
+
+    async fn grant_role_privileges(
+        &self,
+        role_name: String,
+        host_name: String,
+        object: GrantObject,
+        privileges: UserPrivilegeSet,
+        seq: Option<u64>,
+    ) -> Result<Option<u64>> {
+        let role_val_seq = self.get_role(&role_name, seq);
+        let mut role_info = role_val_seq.await?.data;
+        role_info
+            .grants
+            .grant_privileges(&role_name, &host_name, &object, privileges);
+        let seq = self.upsert_role_info(&role_info, seq).await?;
+        Ok(Some(seq))
+    }
+
+    async fn revoke_role_privileges(
+        &self,
+        role_name: String,
+        host_name: String,
+        object: GrantObject,
+        privileges: UserPrivilegeSet,
+        seq: Option<u64>,
+    ) -> Result<Option<u64>> {
+        let role_val_seq = self.get_role(&role_name, seq);
+        let mut role_info = role_val_seq.await?.data;
+        role_info
+            .grants
+            .revoke_privileges(&role_name, &host_name, &object, privileges);
+        let seq = self.upsert_role_info(&role_info, seq).await?;
+        Ok(Some(seq))
+    }
+
+    async fn drop_role(&self, role_name: String, seq: Option<u64>) -> Result<()> {
```

Comment:
It's recommended to use type `MatchSeq` for `seq`.

`Option<_>` is a legacy way to specify which seq to match. It is unclear what to do with `None`: sometimes it means to `add`, sometimes it means to just ignore seq check.